homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,296 @@
+import functools as ft
+import logging
+from typing import Any, ParamSpec, TypeVar, final
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    SERVICE_CLOSE_VALVE,
+    SERVICE_OPEN_VALVE,
+    SERVICE_SET_VALVE_POSITION,
+    SERVICE_STOP_VALVE,
+    SERVICE_TOGGLE,
+    STATE_CLOSED,
+    STATE_CLOSING,
+    STATE_OPEN,
+    STATE_OPENING,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import bind_hass
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "valve"
+SCAN_INTERVAL = timedelta(seconds=15)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+class ValveDeviceClass(StrEnum):
+    """Device class for valve."""
+
+    # Refer to the valve dev docs for device class descriptions
+    WATER = "water"
+    GAS = "gas"
+
+
+DEVICE_CLASSES_SCHEMA = vol.All(vol.Lower, vol.Coerce(ValveDeviceClass))
+
+# DEVICE_CLASS* below are deprecated as of 2021.12
+# use the ValveDeviceClass enum instead.
+DEVICE_CLASSES = [cls.value for cls in ValveDeviceClass]
+DEVICE_CLASS_WATER = ValveDeviceClass.WATER.value
+DEVICE_CLASS_GAS = ValveDeviceClass.GAS.value
+
+# mypy: disallow-any-generics
+
+
+class ValveEntityFeature(IntFlag):
+    """Supported features of the valve entity."""
+
+    OPEN = 1
+    CLOSE = 2
+    SET_POSITION = 4
+    STOP = 8
+
+
+ATTR_CURRENT_POSITION = "current_position"
+ATTR_POSITION = "position"
+
+
+@bind_hass
+def is_closed(hass: HomeAssistant, entity_id: str) -> bool:
+    """Return if the valve is closed based on the statemachine."""
+    return hass.states.is_state(entity_id, STATE_CLOSED)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Track states and offer events for valves."""
+    component = hass.data[DOMAIN] = EntityComponent[ValveEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_OPEN_VALVE, {}, "async_open_valve", [ValveEntityFeature.OPEN]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_CLOSE_VALVE, {}, "async_close_valve", [ValveEntityFeature.CLOSE]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALVE_POSITION,
+        {
+            vol.Required(ATTR_POSITION): vol.All(
+                vol.Coerce(int), vol.Range(min=0, max=100)
+            )
+        },
+        "async_set_valve_position",
+        [ValveEntityFeature.SET_POSITION],
+    )
+
+    component.async_register_entity_service(
+        SERVICE_STOP_VALVE, {}, "async_stop_valve", [ValveEntityFeature.STOP]
+    )
```

Comment:
I don't see how this should be used.

Reply:
For valves that can open half way (not all of them do) there might be a reason to start opening or closing a valve and then stop at whatever the current % is. 

It's not an use case I have myself, but from my understanding of the discussion over in the architecture repo, it was supposed to work as covers do.

Reply:
I left a reaction in at the architectural discussion. Mainly a valve looks like a cover, but stopping a valve does not make sense to me, may be a valve can get `stuck`.

Reply:
I don't find removing it. If @emontnemery can chime in to confirm I'm happy to remove it. There are very few things I like more than deleting code.